client_code/historic/services.py

```diff
@@ -0,0 +1,93 @@
+from anvil_labs import non_blocking
+from anvil_labs.atomic import action, atom, reaction, selector
+
+__version__ = "0.0.1"
+
+
+@atom
+class Archivist:
+    def __init__(self, publisher, projectors=None, deferral=2):
+        """A class to sending events to the server for persistence.
+
+        Parameters
+        ----------
+        publisher : anvil_extras.messaging.Publisher
+        projectors : list
+            of projector names to run after events have been saved
+        deferral : int
+            number of seconds to wait before retrying if a save is already in progress
+        """
+        self.projectors = projectors or []
+        self.deferral = deferral
+        self.saving = []
+        self.pending = []
+        self.failed = []
+        publisher.subscribe(
+            channel="event", subscriber=self, handler=self._handle_message
+        )
+        reaction(lambda: self.pending, self._update_db)
+        self.deferred_save = None
+
+    @property
+    @selector
+    def status(self):
+        pending, saving, failed = self.pending, self.saving, self.failed
+        if failed:
+            return "unsaved"
+        if pending or saving:
+            return "saving"
+        return "saved"
+
+    def _handle_message(self, message):
+        if message.title == "historic.event.occurred":
+            self.pending.append(message.content)
+
+        if message.title == "historic.events.occurred":
+            self.pending.extend(message.content)
+
+    def _handle_result(self, result):
+        """A handler for the result of an async server call
+
+        Parameters
+        ----------
+        result : object
+        """
+        self.saving = []
+
+    @action
+    def _handle_error(self, err):
+        """A handler for errors during an async server call
+
+        Parameters
+        ----------
+        err : Exception
+        """
+        print(f"Error in server call: {err}")
+        self.failed.extend(self.saving)
+        self.saving = []
+
+    def _do_update(self):
+        """Make an async call to the server to save any pending changes"""
+        unsaved = self.pending
+        if not unsaved:
+            return
+        self.saving, self.pending = self.pending, []
+        async_call = non_blocking.call_async(
+            "anvil_labs.historic.save_events", unsaved, projectors=self.projectors
+        )
```

Comment:
does there also need to be a public api for this?
in `client_code/historic/__init__.py`?

Reply:
I don't think so. We fire this off with messages coming in via the publisher

Reply:
just in case we're speaking about different things
I meant

Does there need to be a public `save_events` function?

As in - if I'm using historic in my app should I ever be calling `"anvil_labs.historic.save_events"` myself?

If yes then I think having a `save_events` function in `client_code/historic/__init__.py` is probably the right thing

And this would become

```python
from . import historic

...

    async_call = non_blocking.call_async(historic.save_events, ...)
```

If we don't expect others to be calling `"anvil_labs.historic.save_events"` then I think this is fine

Reply:
I think I'd prefer to deal with the public api separately - it's all synchronous at the moment and I don't really want a mixture of sync and async in there.

How about we put  this in, think about the API and come back to this once we're happy?

Reply:
(I also need to sort out being able to pass the logging level in, so there's already work to do on the public face of it)